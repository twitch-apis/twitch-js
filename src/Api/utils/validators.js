import invariant from 'invariant'
import { conformsTo, defaultsDeep, isBoolean, isString } from 'lodash'

import * as constants from '../constants'

const apiOptions = maybeOptions => {
  /**
   * API options
   * @typedef {Object} ApiOptions
   * @property {string} [clientId] Optional if token is defined.
   * @property {string} [token] Optional if clientId is defined.
   * @property {string} [urlRoot]
   * @property {boolean} [debug=false]
   */
  const shape = {
    debug: isBoolean,
  }

  const shapeWithClientId = {
    ...shape,
    clientId: isString,
  }

  const shapeWithToken = {
    ...shape,
    token: isString,
  }

  const options = defaultsDeep(
    {},
    { ...maybeOptions },
    {
      urlRoot: constants.KRAKEN_URL_ROOT,
      debug: false,
    },
  )

  invariant(
    conformsTo(options, shapeWithClientId) ||
      conformsTo(options, shapeWithToken),
    '[twitch-js/Api] options: Expected valid options',
  )

  return options
}

export { apiOptions }
