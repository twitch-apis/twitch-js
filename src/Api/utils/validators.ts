import invariant from 'invariant'

import conformsTo from 'lodash/conformsTo'
import defaults from 'lodash/defaults'
import isFunction from 'lodash/isFunction'
import isString from 'lodash/isString'
import isUndefined from 'lodash/isUndefined'

import { Options } from '../types'

export const apiOptions = (maybeOptions: any): Options | never => {
  const shape = {
    clientId: (token: unknown) => isUndefined(token) || isString(token),
    token: (token: unknown) => isUndefined(token) || isString(token),
    onAuthenticationFailure: isFunction,
  }

  const options = defaults<Options, Options>(
    { ...maybeOptions },
    {
      clientId: undefined,
      token: undefined,
      onAuthenticationFailure: () => new Promise((_, reject) => reject()),
    },
  )

  invariant(
    conformsTo(options, shape),
    '[twitch-js/Api] options: Expected valid options',
  )

  return options as Options
}
