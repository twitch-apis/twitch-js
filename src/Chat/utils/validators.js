import invariant from 'invariant'

import {
  conformsTo,
  defaults,
  isString,
  isFinite,
  isFunction,
  isBoolean,
  isNil,
} from 'lodash'

import * as constants from '../constants'
import * as sanitizers from './sanitizers'

const chatOptions = maybeOptions => {
  /**
   * Chat options
   * @typedef {Object} ChatOptions
   * @property {string} [username]
   * @property {string} [token] OAuth token (use {@link https://twitchapps.com/tmi/} to generate one)
   * @property {number} [connectionTimeout=CONNECTION_TIMEOUT]
   * @property {number} [joinTimeout=JOIN_TIMEOUT]
   * @property {boolean} [debug=false]
   * @property {Function} [onAuthenticationFailure]
   */
  const shape = {
    username: isString,
    token: value => isNil(value) || isString(value),
    connectionTimeout: isFinite,
    joinTimeout: isFinite,
    debug: isBoolean,
    onAuthenticationFailure: isFunction,
  }

  const options = defaults(
    {
      ...maybeOptions,
      username: sanitizers.username(maybeOptions.username),
    },
    {
      connectionTimeout: constants.CONNECTION_TIMEOUT,
      joinTimeout: constants.JOIN_TIMEOUT,
      debug: false,
      onAuthenticationFailure: () => Promise.reject(),
    },
  )

  invariant(
    conformsTo(options, shape),
    '[twitch-js/Chat] options: Expected valid options',
  )

  return options
}

const clientOptions = maybeOptions => {
  const shape = {
    username: isString,
    password: isString,
    server: isString,
    port: isFinite,
    ssl: isBoolean,
  }

  const options = defaults(
    {
      ...maybeOptions,
      username: sanitizers.username(maybeOptions.username),
      password: sanitizers.password(maybeOptions.token),
    },
    {
      server: constants.CHAT_SERVER,
      port: constants.CHAT_SERVER_SSL_PORT,
      ssl: true,
    },
  )

  invariant(
    conformsTo(options, shape),
    '[twitch-js/Chat/Client] options: Expected valid options',
  )

  return options
}

export { chatOptions, clientOptions }
