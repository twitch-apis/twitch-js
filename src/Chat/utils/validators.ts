import invariant from 'invariant'

import conformsTo from 'lodash/conformsTo'
import defaults from 'lodash/defaults'
import isString from 'lodash/isString'
import isFinite from 'lodash/isFinite'
import isFunction from 'lodash/isFunction'
import isBoolean from 'lodash/isBoolean'
import isNil from 'lodash/isNil'

import * as types from '../types'

import * as constants from '../constants'
import * as sanitizers from './sanitizers'

export const chatOptions = (
  maybeOptions: types.ChatOptions,
): types.ChatOptions => {
  const shape = {
    username: isString,
    token: (value: any) => isNil(value) || isString(value),
    isKnown: isBoolean,
    isVerified: isBoolean,
    connectionTimeout: isFinite,
    joinTimeout: isFinite,
    onAuthenticationFailure: isFunction,
  }

  const options: types.ChatOptions = defaults(
    {
      ...maybeOptions,
      username: sanitizers.username(maybeOptions.username),
      token: sanitizers.token(maybeOptions.token),
    },
    {
      isKnown: false,
      isVerified: false,
      connectionTimeout: constants.CONNECTION_TIMEOUT,
      joinTimeout: constants.JOIN_TIMEOUT,
      onAuthenticationFailure: () => Promise.reject(),
    },
  )

  invariant(
    conformsTo(options, shape),
    '[twitch-js/Chat] options: Expected valid options',
  )

  return options
}

export const clientOptions = (
  maybeOptions: types.ClientOptions,
): types.ClientOptions => {
  const shape = {
    username: isString,
    token: isString,
    server: isString,
    port: isFinite,
    ssl: isBoolean,
    isKnown: isBoolean,
    isVerified: isBoolean,
  }

  const options: types.ClientOptions = defaults(
    {
      ...maybeOptions,
      username: sanitizers.username(maybeOptions.username),
      token: sanitizers.token(maybeOptions.token),
    },
    {
      server: constants.CHAT_SERVER,
      port: constants.CHAT_SERVER_SSL_PORT,
      ssl: true,
      isKnown: false,
      isVerified: false,
    },
  )

  invariant(
    conformsTo(options, shape),
    '[twitch-js/Chat/Client] options: Expected valid options',
  )

  return options
}
