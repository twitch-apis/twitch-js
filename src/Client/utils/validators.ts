import invariant from 'invariant'

import conformsTo from 'lodash/conformsTo'
import defaults from 'lodash/defaults'
import isString from 'lodash/isString'
import isFinite from 'lodash/isFinite'
import isBoolean from 'lodash/isBoolean'

import * as types from '../types'

import * as constants from '../constants'
import * as sanitizers from './sanitizers'

export const clientOptions = (
  options: Partial<types.ClientOptions>,
): types.ClientOptions => {
  const shape = {
    username: isString,
    token: isString,
    server: isString,
    port: isFinite,
    ssl: isBoolean,
    isKnown: isBoolean,
    isVerified: isBoolean,
  }

  const optionsWithDefaults: types.ClientOptions = defaults(
    {
      ...options,
      username: sanitizers.username(options.username),
      token: sanitizers.token(options.token),
    },
    {
      server: constants.CHAT_SERVER,
      port: constants.CHAT_SERVER_SSL_PORT,
      ssl: true,
      isKnown: false,
      isVerified: false,
    },
  )

  invariant(
    conformsTo(optionsWithDefaults, shape),
    '[twitch-js/Chat/Client] options: Expected valid options',
  )

  return optionsWithDefaults
}
