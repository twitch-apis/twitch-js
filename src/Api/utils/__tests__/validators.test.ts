import * as validators from '../validators'

describe('Api/utils/validators', () => {
  const options = { token: 'TOKEN' }

  describe('apiOptions', () => {
    test('should return default chat options', () => {
      const actual = validators.apiOptions(options)

      expect(actual).toMatchSnapshot()
    })

    test('should throw if token is missing', async () => {
      expect(() =>
        validators.apiOptions({ ...options, token: undefined }),
      ).toThrow()
    })
  })
})
