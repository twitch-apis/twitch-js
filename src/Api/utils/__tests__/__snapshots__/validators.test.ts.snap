// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api/utils/validators apiOptions should return default chat options 1`] = `
Object {
  "clientId": undefined,
  "onAuthenticationFailure": undefined,
  "token": "TOKEN",
}
`;
